File: 19. Remove Nth Node From End of List.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeNthFromEnd(head *ListNode, n int) *ListNode {
    length := findLength(head)
    if length == n {
        return head.Next
    }
    length = length - n - 1
    pointer := head
    for length > 0 {
        pointer = pointer.Next
        length--
    }
    pointer.Next = pointer.Next.Next
    return head
}
func findLength(pointer *ListNode) int {
    if pointer == nil {
        return 0
    }
    return 1 + findLength(pointer.Next)
}

File: java/19. Remove Nth Node From End of List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode reversed = reverse(head);
        if (n == 1) return reverse(reversed.next);
        ListNode pointer = reversed;
        while (n-- > 2) {
            pointer = pointer.next;
        }
        pointer.next = pointer.next.next;
        return reverse(reversed);
    }
    private ListNode reverse(ListNode node) {
        ListNode previous = null;
        while (node != null) {
            ListNode temp = node.next;
            node.next = previous;
            previous = node;
            node = temp;
        }
        return previous;
    }
}